### ISSUE TITLE
Leveled fields: Fields present only at certain log levels

### ISSUE THREAD

**Issue by abhinav** [uber-go/zap#1078]
An issue that users sometimes have is this:

They have a log statement like the following,

```
logger.Info("processing request", zap.String("fromUser", fromUser))
```

They would like to include additional information in that log statement, but only if the debug level is enabled.

Currently, the options are:

1. Copy paste the log statement

```
if debugLevel {
  logger.Info("processing request", zap.String("fromUser", fromUser), zap.Object("request", request))
} else {
  logger.Info("processing request", zap.String("fromUser", fromUser))
}
```

2. Add another log statement

```
logger.Info("processing request", zap.String("fromUser", fromUser))
logger.Debug("processing request", zap.Object("request", request))
```

3. custom helpers like the following:

```
func levelField(log *zap.Logger, lvl zap.Level, field zap.Field) zap.Field {
  if !log.Core().Enabled(lvl) { return zap.Skip() }
  return field
}
```

```
logger.Info("processing request",
  zap.String("fromUser", fromUser),
  levelField(logger, zap.DebugLevel, zap.Object("request", request)))
```

I think it might be valuable for Zap to natively provide some concept of "leveled fields"—fields present only at certain log levels—based on something like the levelField helper above.

I don't have an exact design in mind, but Zap-native support for something like this would likely be able to drop that extra `log` argument.

**Comment by abhinav** [comment]
Summarizing discussion with @mway here:

Matt proposed the following options:

1. Add options to existing field constructors.

    ```go
    func String(key, value string, opts ...FieldOption)
    
    type FieldOption
      func FieldLevel(Level) FieldOption
    ```

2. Parallel constructors.

    ```go
    func String(key, value string)
    func StringLevel(key, value string, lvl Level)
    ```

3. Method on zapcore.Field.

    ```go
    package zapcore
    
    type Field struct{ ... }
        func (Field) IfLevel(Level) Field
    ```

4. Name-based filtering at the logger level:

    ```go
    func WithFieldLevelFilters(map[string]Level) Option
    ```

---

My thoughts on the options:

1. Breaking change. I would like to avoid that for such a small feature.
2. Too much noise. It'll double the number of field constructors.
3. I like this. Maybe we can come up with a better name.
4. Stringly-typed API. Would prefer to avoid. Requires field name to be stable/unique across the application.

So *if* we were going to solve this problem, based on current options, I lean (3).
I'm also okay with saying that this isn't something Zap should support natively at this time,
and that it's okay for users to have to write the helper function I suggested above.

**Comment by prashantv** [comment]
I think this is a useful feature, and more generally have found cases where the fields to log depend on some condition, and I find myself writing code like:
```go
fieldsToLog := []zap.Field{...}
if someCondition {
  fieldsToLog = append(fieldsToLog, zap.String(...), zap.Int(...))
}
```

Looking at the options, agree on (1), (2), and (4) with @abhinav.

Re: 3: This is an interesting idea, and could be used for more generic helpers, like `func (Field) If(condition bool) Field` as well. The chaining approach does feel like a builder which we don't see too frequently in Go.

Another variant (more similar to the `levelIfField` is to have a top-level function: `func IfLevel(level Level, fields... Field) Field`, which may be a little more verbose, but it also allows providing multiple fields at once, so if there's more than one field being added, it'll likely be less noise.

We could take that idea a little further, and have field constructors per-level, e.g., `func DebugField(Field) Field` (could also take multiple fields), which might be good for readability too:
```go
logger.Info("my message",
  zap.String("foo", "bar"),
  zap.DebugField(zap.String("debugOnly", "baz")),
)
```

**Comment by abhinav** [comment]
Oh, I like the idea of the top-level functions. Although I suspect that accepting multiple fields there will make that a one-alloc-each operation because we'll need to be able to turn `[]Field` into a `Field` with `zap.Inline`—which may or may not be a dealbreaker here.

**Comment by prashantv** [comment]
True, may want to separate the single-field and multi-field functions.

**Comment by mway** [comment]
I think we should start with single-field filtering and extend to `[]Field` if we see a compelling need. I don't personally think it's worth the complexity/overhead to support a variable number fields for callsite filtering, at least not initially:

```go
logger.Info(
  "handled request",
  zap.String("caller", "..."),
  zap.String("method", "..."),
  zap.DebugField(zap.String("debug1", "...")),
  zap.DebugField(zap.String("debug2", "...")),
  zap.DebugField(zap.String("debug3", "...")),
)
```

is arguably not worse than

```go
logger.Info(
  "handled request",
  zap.String("caller", "..."),
  zap.String("method", "..."),
  zap.DebugFields(
    zap.String("debug1", "..."),
    zap.String("debug2", "..."),
    zap.String("debug3", "..."),
  ),
)
```

IMO, if we want to deal with `[]Field` as a semantic concept (i.e., "this subset of message fields relate to each other in some special way compared to the rest of the fields"), it might be better to formalize "field groups" as a concept first (if we feel strongly about discrete groups of fields), and then subsequently add support for group-level filtering.

We should probably also be clear about whether *all* levels add value, or only some. I can't think of a useful case in which a `PanicField`, `FatalField`, `ErrorField`, or even `WarnField` would be useful, but maybe such a case does exist. Even so, it's probably worth starting with `DebugField` (and *maybe* `InfoField`, if we're confident enough) and feeling it out first.

It may also be useful (if possible) to prototype the API in an experimental package (e.g. `go.uber.org/zap/x/filter`) first. If we can do that, and if we're not sold on APIs and/or the usefulness of filtering at some log levels, then perhaps a more general version would be good to start (e.g. `filter.Field(zapcore.Level, zap.Field) zap.Field`) so we can gather data/use cases before committing.

**Comment by ozfive** [comment]
> I think we should start with single-field filtering and extend to `[]Field` if we see a compelling need. I don't personally think it's worth the complexity/overhead to support a variable number fields for callsite filtering, at least not initially:
> 
> ```go
> logger.Info(
>   "handled request",
>   zap.String("caller", "..."),
>   zap.String("method", "..."),
>   zap.DebugField(zap.String("debug1", "...")),
>   zap.DebugField(zap.String("debug2", "...")),
>   zap.DebugField(zap.String("debug3", "...")),
> )
> ```
> 
> is arguably not worse than
> 
> ```go
> logger.Info(
>   "handled request",
>   zap.String("caller", "..."),
>   zap.String("method", "..."),
>   zap.DebugFields(
>     zap.String("debug1", "..."),
>     zap.String("debug2", "..."),
>     zap.String("debug3", "..."),
>   ),
> )
> ```
> 
> IMO, if we want to deal with `[]Field` as a semantic concept (i.e., "this subset of message fields relate to each other in some special way compared to the rest of the fields"), it might be better to formalize "field groups" as a concept first (if we feel strongly about discrete groups of fields), and then subsequently add support for group-level filtering.
> 
> We should probably also be clear about whether _all_ levels add value, or only some. I can't think of a useful case in which a `PanicField`, `FatalField`, `ErrorField`, or even `WarnField` would be useful, but maybe such a case does exist. Even so, it's probably worth starting with `DebugField` (and _maybe_ `InfoField`, if we're confident enough) and feeling it out first.
> 
> It may also be useful (if possible) to prototype the API in an experimental package (e.g. `go.uber.org/zap/x/filter`) first. If we can do that, and if we're not sold on APIs and/or the usefulness of filtering at some log levels, then perhaps a more general version would be good to start (e.g. `filter.Field(zapcore.Level, zap.Field) zap.Field`) so we can gather data/use cases before committing.

If you do decide to add support for field groups in the future would you consider adding group-level filtering as well? Group level filtering would be useful in situations where you have multiple related fields that you only want to log at a certain log level. For example, let's say you have a set of fields related to HTTP requests and you only want to log them at the Debug level. You would define a "request" group that includes all of these fields and then filter the group based on log level.

So instead of:
```go
logger.Debug("processing request",
  zap.String("url", req.URL.String()),
  zap.String("method", req.Method),
  zap.String("user-agent", req.UserAgent()),
  zap.Int("content-length", int(req.ContentLength)),
  zap.String("remote-addr", req.RemoteAddr),
  zap.String("proto", req.Proto),
)
```

You would do:
```go
reqGroup := zap.Group("request", zap.String("url", req.URL.String()), zap.String("method", req.Method), zap.String("user-agent", req.UserAgent()), zap.Int("content-length", int(req.ContentLength)), zap.String("remote-addr", req.RemoteAddr), zap.String("proto", req.Proto))

logger.Debug("processing request",
  zap.Fields(reqGroup.Filter(zap.DebugLevel)),
)
```

This way, if you decide to add more fields to the "request" group later on, you can filter them all together without having to add individual filters for each field.

**Comment by ChannyClaus** [comment]
took a stab at this https://github.com/uber-go/zap/pull/1251 - let me know if this is what you guys had in mind!
